Plutarch/TryFrom.hs

```diff
@@ -0,0 +1,404 @@
+  , PTryFrom PData (PAsData b)
+  ) =>
+  PTryFrom PData (PAsData (PBuiltinMap a b))
+  where
+  type PTryFromExcess PData (PAsData (PBuiltinMap a b)) = PBuiltinList (PBuiltinPair (PAsData a) (PAsData b))
+  ptryFrom opq = runTermCont $ do
+    verMap <- tcont $ plet (pasMap # opq)
+    -- I've not obtained a `PBuiltinList (PBuiltinPair a b)`
+    let verifyPair :: Term _ (PBuiltinPair PData PData :--> PBuiltinPair (PAsData a) (PAsData b))
+        verifyPair = plam $ \tup -> unTermCont $ do
+          (verfst, _) <- TermCont $ ptryFrom @PData @(PAsData a) $ pfstBuiltin # tup
+          (versnd, _) <- TermCont $ ptryFrom @PData @(PAsData b) $ psndBuiltin # tup
+          pure $ ppairDataBuiltin # verfst # versnd
+    ver <- tcont $ plet $ pmap # verifyPair # verMap
+    pure (punsafeCoerce opq, ver)
+
+instance {-# OVERLAPPING #-} PTryFrom PData (PAsData (PBuiltinMap POpaque POpaque)) where
+  type PTryFromExcess PData (PAsData (PBuiltinMap POpaque POpaque)) = PBuiltinList (PBuiltinPair (PAsData POpaque) (PAsData POpaque))
+  ptryFrom opq = runTermCont $ do
+    ver <- tcont $ plet (pasMap # opq)
+    pure $ (punsafeCoerce opq, punsafeCoerce ver) -- PAsData POpaque ~ PData
+
+instance
+  {-# OVERLAPPABLE #-}
+  ( PTryFrom PData a
+  , a ~ PAsData b
+  , PIsData b
+  ) =>
+  PTryFrom PData (PAsData (PBuiltinList a))
+  where
+  type PTryFromExcess PData (PAsData (PBuiltinList a)) = PBuiltinList a
+  ptryFrom opq = runTermCont $ do
+    let lst :: Term _ (PBuiltinList a)
+        lst = punsafeBuiltin PLC.UnListData # opq
+        verify :: a ~ PAsData b => Term _ (PAsData b :--> a)
+        verify = plam $ \e ->
+          unTermCont $ do
+            (wrapped, _) <- TermCont $ ptryFrom @PData @a $ pforgetData e
+            pure wrapped
+    ver <- tcont $ plet $ pmap # verify # lst
+    pure $ (punsafeCoerce opq, ver)
+
+instance
+  {-# OVERLAPPABLE #-}
+  ( PTryFrom PData a
+  , a ~ PAsData a'
+  , PIsData a'
+  , PTryFrom PData b
+  , b ~ PAsData b'
+  , PIsData b'
+  ) =>
+  PTryFrom PData (PAsData (PBuiltinPair a b))
+  where
+  type PTryFromExcess PData (PAsData (PBuiltinPair a b)) = PBuiltinPair a b
+  ptryFrom opq = runTermCont $ do
+    tup <- tcont $ plet (pfromData $ punsafeCoerce opq)
+    let fst' :: Term _ a
+        fst' = unTermCont $ fst <$> TermCont (ptryFrom @PData @a $ pforgetData $ pfstBuiltin # tup)
+        snd' :: Term _ b
+        snd' = unTermCont $ fst <$> TermCont (ptryFrom @PData @b $ pforgetData $ psndBuiltin # tup)
+    ver <- tcont $ plet $ ppairDataBuiltin # fst' # snd'
+    pure $ (punsafeCoerce opq, ver)
+
+instance
+  {-# OVERLAPPABLE #-}
+  ( PTryFrom PData (PAsData b)
+  , PTryFrom PData (PAsData (PDataRecord xs))
+  , x ~ (s ':= b)
+  , PIsData (PDataRecord xs)
+  , PTryFromExcess PData (PAsData (PDataRecord xs)) ~ PDataRecord xs
+  ) =>
+  PTryFrom PData (PAsData (PDataRecord (x ': xs)))
+  where
+  type PTryFromExcess PData (PAsData (PDataRecord (x ': xs))) = PDataRecord (x ': xs)
```

Comment:
so an issue here is that we're actually discarding the "excess" of the inner verification. When you make it such that `PTryFromExcess a b s` is returned instead, you can return a Haskell pair of the excess stuff, such that everything is included.